### PR TITLE
[SPARK-29986][SQL] casting string to date/timestamp/interval should trim all whitespaces

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

A java like string trim method trims all whitespaces that less or equal than 0x20. currently, our UTF8String handle the space =0x20 ONLY. This is not suitable for many cases in Spark, like trim for interval strings, date, timestamps, PostgreSQL like cast string to boolean.

### Why are the changes needed?

improve the white spaces handling in UTF8String, also with some bugs fixed

### Does this PR introduce any user-facing change?

yes,
string with `control character` at either end can be convert to date/timestamp and interval now

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @wangyum @HyukjinKwon thanks in advance.

**Comment by cloud-fan** [comment]
The thing we care is the SQL behavior, not java behavior. Can you check other databases and see if the `trim` function trims all chars whose ascii code <= 0x20?

**Comment by yaooqinn** [comment]
> The thing we care is the SQL behavior, not java behavior. Can you check other databases and see if the `trim` function trims all chars whose ascii code <= 0x20?

Not changing string functions, just for other types listed in pr desc.
Besides, the former interval parser should be able to handle these control chars

```sql
postgres=# select date E'2019-01-01\t';
    date
------------
 2019-01-01
(1 row)

postgres=# select date E'2019-01-01\t';
    date
------------
 2019-01-01
(1 row)

postgres=# select cast(E'1\t' as boolean);
 bool
------
 t
(1 row)

postgres=# select timestamp E'2019-01-01\t';
      timestamp
---------------------
 2019-01-01 00:00:00
(1 row)
```

#### hive
```sql
select date('2019-10-10\t');
"_c0"
"2019-10-10"

```

**Comment by cloud-fan** [comment]
can you try `select trim('2019-10-10\t')` in other databases?

**Comment by yaooqinn** [comment]
> can you try `select trim('2019-10-10\t')` in other databases?

we are not changing string trim here, but here is the test result.
```sql
postgres=# select length(trim(E'2019-10-10\t'));
 length
--------
     11
(1 row)
```

```
presto> select length(trim('2019-10-10' || CHR(09)));
 _col0
-------
    10
(1 row)
```
```
hive
select length(trim('2019-10-10\t')) 
"_c0"
"11"
```

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -553,6 +553,29 @@ public UTF8String trim() {
     return copyUTF8String(s, e);
   }
 
+  /**
+   * Trims whitespaces (<= ASCII 32) from both ends of this string.
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Interesting, so SQL trim just removes space, and java.lang.String.trim() removes everything <= 32. Maybe we could refer to that in this doc, that this is the purpose of this additional trim method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala [hunk 2]
```diff
@@ -42,7 +42,7 @@ case class PostgreCastToBoolean(child: Expression, timeZoneId: Option[String])
   override def castToBoolean(from: DataType): Any => Any = from match {
     case StringType =>
       buildCast[UTF8String](_, str => {
-        val s = str.trim().toLowerCase()
+        val s = str.trimAll().toLowerCase()
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala]: Are these really the only cases that should trim this way?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -553,6 +553,29 @@ public UTF8String trim() {
...
+   *
+   * @return this string with no spaces at the start or end
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Nit: not just space

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala]: the only left one is in string trim function, which I checked pg and hive are spaces trim only.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala]: presto string trim does java like way.

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala]: Hm, it just seems odd to trim this just for booleans and dates?  is there more context here?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastToBoolean.scala]: for integral types, we currently handle relative issue here https://github.com/apache/spark/pull/26622

**Comment by yaooqinn** [comment]
```scala
scala> " 234\t ".toDouble
res0: Double = 234.0

scala> java.lang.Double.valueOf(" 234\t ")
res1: Double = 234.0
```
the way spark deal with casting string to approximate numeric - float and double also trim control characters silently. @cloud-fan
@srowen

I take a look at SQL standard about string trim
> 3) If trim character is not specified, then ' ' is implicit.

>b) If <trim character> is specified, then let SC be the value of trim character; otherwise, let SC be
space.

**Comment by yaooqinn** [comment]
I also look up casting string to other types, the standard only says 

> If SD is character string, then SV is replaced by TRIM ( BOTH ' ' FROM VE )

It seem that most modern dbs are varying this rule.

**Comment by cloud-fan** [comment]
From SQL standard
```
let SRC be <trim source>. TRIM ( SRC ) is equivalent to TRIM ( BOTH ' ' FROM SRC ).
```
```
cast specification
If SD is character string, then SV is replaced by SV with any leading or trailing <space>s removed.
```
some related information
```
L( <left bracket> <colon> SPACE <colon> <right bracket> )
is the set of all character strings of length 1 (one) that are the <space> character.
r) L( <left bracket> <colon> WHITESPACE <colon> <right bracket> )
is the set of all character strings of length 1 (one) that are white space characters.
...
white space
consecutive sequences of one or more characters that have no glyphs
```

So space means `' '`, and white space means all chars whose ascii code <= 32. `trim` and `cast` should only remove spaces.

However, seems most of the DBs don't follow the cast part, and we rely on `Double.valueOf` so hard to change this behavior. I think it's OK to trim white spaces in cast.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 1 above)
Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: yea, let's mention it's the same as ` java.lang.String.trim`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -553,6 +553,29 @@ public UTF8String trim() {
...
+      return EMPTY_UTF8;
+    }
+    // skip all of the whitespaces (<=0x20) in the right side
+    int e = this.numBytes - 1;
+    while (e > s && getByte(e) <= 0x20) e--;
+    if (s == 0 && e == numBytes - 1) {
+      // Nothing trimmed
+      return this;
+    }
+    return copyUTF8String(s, e);
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: looking at the caller side, I think it's safe to not copy the data. We can add a caveat in the javadoc that: this method doesn't copy the data and the caller side should do copy themselves if they want to hold it for a while.

Comment (yaooqinn) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: do we not copy when it is an `EMPTY_UTF8` either? A bit odd if we do it differently.

Comment (cloud-fan) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: ah that's a good point. OK let's leave it.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -553,6 +553,34 @@ public UTF8String trim() {
     return copyUTF8String(s, e);
   }
 
+  /**
+   * Trims whitespaces (<= ASCII 32) from both ends of this string.
+   *
+   * Note that, this method is the same as java's {@link String#trim}, and different from
+   * {@link UTF8String#trim()} which only remove only spaces(= ASCII 32) from both ends.
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: nit: remove one "only"

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -553,6 +553,34 @@ public UTF8String trim() {
...
+   *
+   * @return A UTF8String whose value is this UTF8String, with any leading and trailing white
+   * space removed, or this UTF8String if it has no leading or trailing whitespace.
+   *
+   */
+  public UTF8String trimAll() {
+    int s = 0;
+    // skip all of the whitespaces (<=0x20) in the left side
+    while (s < this.numBytes && getByte(s) <= 0x20) s++;
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: shall we use `' '` instead of `0x20`?

Comment (yaooqinn) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: maybe also replace them in method trim? I don't know if whether ok or not I do so in this pr

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, can we add a migration guide?

**Comment by yaooqinn** [comment]
migration guide added

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by yaooqinn** [comment]
Thanks for merging and the more suitable PR title :)

**Pull request closed without merging**